.github/workflows/qpy.yml

```diff
@@ -0,0 +1,39 @@
+name: QPY
+
+on:
+  push:
+    branches:
+      - 'main'
+      - 'stable/*'
+  pull_request:
+  merge_group:
+concurrency:
+  group: ${{ github.repository }}-${{ github.ref }}-${{ github.head_ref }}-${{ github.workflow }}
+  cancel-in-progress: ${{ github.event_name == 'pull_request' }}
+
+jobs:
+  backward_compat:
+    if: github.repository_owner == 'Qiskit'
+    name: Backwards compatibility
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses: actions/setup-python@v5
+        with:
+          python-version: '3.9'
+
+      - uses: dtolnay/rust-toolchain@stable
+
+      - uses: actions/cache@v4
+        with:
+          path: test/qpy_compat/qpy_cache
+          # The hashing is this key can be too eager to invalidate the cache,
+          # but since we risk the QPY tests failing to update if they're not in
+          # sync, it's better safe than sorry.
+          key: qpy-${{ hashFiles('test/qpy_compat/**') }}
```

Comment:
Will this hash only on the files in the checkout from git, or will it include all the qpy files we generate during the run? I can't remember if the hashing is only performed at the start of the job or the end of the job.

Reply:
If I got this right (I _think_ I did), the hash is calculated only once on the initial git checkout (so doesn't include the QPY files), but when pushing back to the cache, it _will_ include the generated files, so subsequent lookups will retrieve them.  It's important that the QPY files aren't part of the hash key because they're not fully deterministic - they include randomly generated UUID payloads in some of the parameters.

Reply:
Ok, yeah that's why I was asking, especially because the docs aren't clear: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#hashfiles but we can keep an eye on it post merge and adjust if it's a problem. The only option I can think of is manually listing out the files in the tree we care about tying to the cache.

Reply:
Back before I opened this PR I checked it in a bunch of configurations on my fork, and the caching all seemed to be working the way I expected/hoped.